### PR TITLE
Cross region support for CRT Client

### PR DESCRIPTION
**Pull request by joviegas**
## Motivation and Context

Cross region support for CRT Client

## Modifications

- Api for crossRegionEnabled in Crt client
- Pass AwsSigning Config to crt if SigningRegion is changed

## Testing
- Integ test for get/put/copy.
## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/crt/DefaultS3CrtAsyncClientTest.java [hunk 1]
```diff
@@ -92,4 +94,23 @@ void invalidConfig_shouldThrowException(long value) {
...
+
+    @Test
+    void crtCrtClient_with_crossRegionAccessEnabled_asTrue(){
+        S3AsyncClient crossRegionCrtClient = S3AsyncClient.crtBuilder().crossRegionAccessEnabled(true).build();
+        assertThat(crossRegionCrtClient).isInstanceOf(DefaultS3CrtAsyncClient.class);
+        assertThat(((DelegatingS3AsyncClient)crossRegionCrtClient).delegate()).isInstanceOf(S3CrossRegionAsyncClient.class);
+    }
+
+    @Test
+    void crtCrtClient_with_crossRegionAccessEnabled_asFalse(){
```
Comment (davidh44) [inline services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/crt/DefaultS3CrtAsyncClientTest.java:106]: nit: just one "crt" here and above?

File: services/s3/src/it/java/software/amazon/awssdk/services/s3/crt/TestResponseTransformer.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.CompletableFuture;
+import software.amazon.awssdk.core.async.AsyncResponseTransformer;
+import software.amazon.awssdk.core.async.SdkPublisher;
+import software.amazon.awssdk.http.async.SimpleSubscriber;
+import software.amazon.awssdk.services.s3.model.GetObjectResponse;
+
+public final class TestResponseTransformer implements AsyncResponseTransformer<GetObjectResponse, Void> {
```
Comment (davidh44) [inline services/s3/src/it/java/software/amazon/awssdk/services/s3/crt/TestResponseTransformer.java:27]: what is the use case for this class?

Comment (joviegas) [reply services/s3/src/it/java/software/amazon/awssdk/services/s3/crt/TestResponseTransformer.java:27]: Will remove this

File: services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/crt/DefaultS3CrtAsyncClientTest.java (hunk 1 above)
Comment (joviegas) [reply services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/crt/DefaultS3CrtAsyncClientTest.java]: done

**Review by davidh44 — APPROVED** [review]

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtAsyncHttpClient.java [hunk 3]
```diff
@@ -130,6 +134,13 @@ public CompletableFuture<Void> execute(AsyncExecuteRequest asyncRequest) {
             .withResponseHandler(responseHandler)
             .withResumeToken(resumeToken);
 
+        // Create a new SigningConfig object only if the signing region has changed from the previously configured region.
+        if (signingRegion != null && !s3ClientOptions.getRegion().equals(signingRegion)) {
```
Comment (dagnir) [inline services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtAsyncHttpClient.java:138]: `s3ClientOptions.getRegion().equals(signingRegion)` is comparing different types, `String` and `Region`. Should be using `signingRegion.id()`

File: services/s3/src/it/java/software/amazon/awssdk/services/s3/crt/S3CrossRegionCrtIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+import software.amazon.awssdk.services.s3.S3IntegrationTestBase;
+import software.amazon.awssdk.services.s3.model.CopyObjectResponse;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.services.s3.model.GetObjectResponse;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.S3Exception;
+import software.amazon.awssdk.testutils.RandomTempFile;
+import software.amazon.awssdk.testutils.service.AwsTestBase;
+
+public class S3CrossRegionCrtIntegrationTest extends S3IntegrationTestBase {
```
Comment (dagnir) [inline services/s3/src/it/java/software/amazon/awssdk/services/s3/crt/S3CrossRegionCrtIntegrationTest.java:54]: Should this implement the same tests from #4128?

File: services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtAsyncHttpClient.java (hunk 3 above)
Comment (joviegas) [reply services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtAsyncHttpClient.java]: Good catch !

File: services/s3/src/it/java/software/amazon/awssdk/services/s3/crt/S3CrossRegionCrtIntegrationTest.java (hunk 4 above)
Comment (joviegas) [reply services/s3/src/it/java/software/amazon/awssdk/services/s3/crt/S3CrossRegionCrtIntegrationTest.java:54]: I added a separate test class for Crt to test API response integrity in 4128 . In 4128 test cases it covers generic API responses. This `S3CrossRegionCrtIntegrationTest` is specifically for testing CRT's API functionalities, such as copyObject. I believe it's important to keep this test case as it verifies the combination of Decorated CRT and Decorated Cross region integrated with CRT library and S3. Thus not removing this.

Comment (dagnir) [reply services/s3/src/it/java/software/amazon/awssdk/services/s3/crt/S3CrossRegionCrtIntegrationTest.java:54]: okay that sounds good, but will it also implement those tests?

Comment (joviegas) [reply services/s3/src/it/java/software/amazon/awssdk/services/s3/crt/S3CrossRegionCrtIntegrationTest.java:54]: Yeap , https://github.com/aws/aws-sdk-java-v2/pull/4128 already has those changes

**Review by dagnir — APPROVED** [review]

**Pull request merged**